### ISSUE TITLE
Third-party command discovery

### ISSUE THREAD

**Issue by schlessera** [wp-cli/ideas#51]
As the [package index has now been deprecated](https://make.wordpress.org/cli/2017/07/18/feature-development-discussion-recap/), we should start the discussion about whether we want a new discovery mechanism, and how that could work.

As the installation of packages has now been split off of the discovery mechanism, discovery can just as well be something entirely external to WP-CLI. It could be a full-blown package repository, but it could just as well be a specialized search engine.

**Comment by danielbachhuber** [comment]
What about https://github.com/search?q=topic%3Awp-cli-package+org%3Awp-cli&type=Repositories ?

**Comment by schlessera** [comment]
That will only cover commands stored on GitHub. Ideally, we'd have a discovery mechanism for finding both stand-alone commands as well as commands included with plugins, no matter where the source files are stored.

**Comment by danielbachhuber** [comment]
> as well as commands included with plugins

Oh, this is an interesting idea.

**Comment by danielbachhuber** [comment]
> > as well as commands included with plugins
>
> Oh, this is an interesting idea.

For plugin commands, maybe the more appropriate place would be the WordPress.org plugin repo. It'd be cool to work out a way for plugin authors to easily incorporate that detail.

**Comment by schlessera** [comment]
Yes, that might be feasible. Add a meta tag to the plugin file header, with something like:
```
commands: event list, event get, ...
```

Another approach would be to scan the plugin files for `WP_CLI::add_command()` calls.

**Comment by miya0001** [comment]
> commands: event list, event get, ...

Oh, nice. 👍

**Comment by schlessera** [comment]
Related: https://meta.trac.wordpress.org/ticket/3126

**Comment by nyl2002** [comment]
I have found https://vimawesome.com/ to be helpful for Vim plugins.

I find their list design to be both compact and also clear, with the added benefit of color-coded categories.
I also think that vim and wp-cli are a natural match and could help promote each other.

It would be useful to both have more WordPress vim packages (https://vimawesome.com/?q=wordpress) and also to let wp-cli profit from features in vim plugins such as UltiSnips, fuzzy finder, git integration, etc.

A WordPress directory plugin for NERD tree might be helpful.

I created an issue over there:
https://github.com/vim-awesome/vim-awesome/issues/381

**Comment by stefanos82** [comment]
Maybe https://wpackagist.org/ is much more appropriate suggestion than [vim-awesome](https://vimawesome.com).

**Comment by szepeviktor** [comment]
How about Packagist?

https://packagist.org/packages/wp-cli/core-command?query=&type=wp-cli-package

**Comment by redtux** [comment]
Hi, I just found out about the fact that `wp package` is deprecated by running `wp package browse`.

Would it be possible to include this information in the docs? https://developer.wordpress.org/cli/commands/package/

> Although the package index will remain in place for backward compatibility reasons, it has been deprecated and will not be updated further. Please refer to https://github.com/wp-cli/ideas/issues/51 to read about its potential replacement.

**Comment by gilzow** [comment]
Was a replacement ever decided on? OR if not, can we resurface the discussion?

**Comment by danielbachhuber** [comment]
> Was a replacement ever decided on?

It was not.

> OR if not, can we resurface the discussion?

Sure, how would you suggest moving the discussion forward?

**Comment by gilzow** [comment]
> Sure, how would you suggest moving the discussion forward?

Thanks for the response, @danielbachhuber ! 

There are several good suggestions above, but I've been unable to find (so far) why none of them were selected as appropriate. I've [read the blog post](https://make.wordpress.org/cli/2017/07/18/feature-development-discussion-recap/) and it seems that a packagist-like solution (Satis, wpackagist, etc) would solve both the central search location, and discoverability issues.  

They do **not** address discovering cli packages that are bundled into plugins/themes; _HOWEVER_, one could argue in those situations the cli functionality is worthless without the plugin/theme it is bundled into. 

IMO, I don't think plugin/theme cli discoverability should be a road-block for a centralized wp-cli package repository. I'm curious why none of the above solutions were ever implemented. Resources? Sponsor? Agreement?  

**Edit**: mispelling

**Comment by danielbachhuber** [comment]
> I'm curious why none of the above solutions were ever implemented. Resources? Sponsor? Agreement?

Honestly, I think it's more a lack of momentum than anything else. It's a little bit of a solution in search of a problem — in reality, there aren't a ton of new packages created on a regular basis.

At this point, my preferred approach would be:
1. A page in the [handbook](https://make.wordpress.org/cli/handbook/) that lists community packages maintained outside of the WP-CLI community.
2. Create more repos inside of this GitHub organization for common, general purpose commands.

Out of curiosity, how did you stumble upon this issue? What problem are you hoping to see addressed?

**Comment by gilzow** [comment]
> Out of curiosity, how did you stumble upon this issue? What problem are you hoping to see addressed?

I've been writing several wp-cli packages that I've been installing via git url. I'm now doing a presentation on writing custom wp-cli packages and noticed the [handbook](https://make.wordpress.org/cli/handbook/guides/commands-cookbook/#overview) [mentions](https://make.wordpress.org/cli/handbook/guides/commands-cookbook/#command-types) the package index [numerous](https://make.wordpress.org/cli/handbook/references/tools/) times. But when I went to investigate how to submit a package to the index (linked from the package-index [page](https://wp-cli.org/package-index/)), discovered it was [deprecated](https://github.com/wp-cli/package-index).  That led me down a rabbit hole that ultimately brought me here. 😄  

> there aren't a ton of new packages created on a regular basis.

I can't speak with confidence on whether there are, or there aren't, but I'll offer this counterpoint: one possibility is people follow the same path I did above, see that the index is deprecated, but then just stop. As an example of this, you can read the discussion surrounding the deprecation (and reversal) of [Gitlab's public accessibility job template](https://gitlab.com/gitlab-org/gitlab/-/issues/390424).

**Comment by danielbachhuber** [comment]
Thanks for the explanation 😊

> I've been writing several wp-cli packages that I've been installing via git url.

Are these packages meant for public consumption? Do you intend to maintain and support them?

> I can't speak with confidence on whether there are, or there aren't, but I'll offer this counterpoint: one possibility is people follow the same path I did above, see that the index is deprecated, but then just stop. 

Sure, makes sense. If there's interest at an upcoming Contributor Day, we could potentially discuss reviving a package index cc @wp-cli/maintainers

**Comment by swissspidy** [comment]
A curated list of some useful and popular commands sounds like a good start in the interim.

Highlighting plugins offering commands would also be nice. Maybe for now also in a curated way? There are around 600 in the repo: https://wpdirectory.net/search/01H46J87X60A5XC4KSJ1A42QVH

**Comment by gilzow** [comment]
> Are these packages meant for public consumption? 

Possibly? I could certainly make them more generic to see if anyone else can use them. So far they've mostly been for my own use. 

> Do you intend to maintain and support them?

Yes, if I end up putting them in a public directory of some sorts, I'd definitely continue to maintain them.

**Comment by schlessera** [comment]
@swissspidy I already created an "awesome list" in the past: https://github.com/schlessera/awesome-wp-cli

**Comment by gilzow** [comment]
> @swissspidy I already created an "awesome list" in the past: https://github.com/schlessera/awesome-wp-cli

Is it ok if I promote/publicize that repository?

**Comment by ahathaway** [comment]
Wondering if there has been any progress/discussion on the possibility of getting the package manager for wp-cli working again? There are those of us who manage hundreds of wordpress websites and having a working, dependable package manager for wp cli would be very nice. 

What is everyone else doing instead?

**Comment by swissspidy** [comment]
What is not working for you?

Discovering & installing WP-CLI packages from Packagist should be working as usual.
